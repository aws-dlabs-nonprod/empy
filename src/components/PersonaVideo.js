import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import * as ActionTypes from '@constants/ActionTypes';
import { CSSTransition } from 'react-transition-group';
import Media from '@style/media';

import { SoulMachinesContext } from '@contexts/SoulMachines';

class PersonaVideo extends PureComponent {

    static contextType = SoulMachinesContext;

    constructor(props, context) {
        super(props, context);

        this.videoRef = React.createRef();

        this.handleResize = this.handleResize.bind(this);
        this.handleVideoPlaying = this.handleVideoPlaying.bind(this);

        this.state = {
            videoStarted: false
        };
    }

    componentDidMount() {
        this.updateVideoStream();

        window.addEventListener('resize', this.handleResize);
    }

    componentWillUnmount() {
        window.removeEventListener('resize', this.handleResize);
    }

    componentDidUpdate(prevProps) {
        const { isConnected } = this.props;

        this.updateVideoStream();

        if (!isConnected) {
            this.setState({ videoStarted: false });
        }
    }

    handleResize() {
        const videoEl = this.videoRef.current;

        if (videoEl) {
            this.props.setVideoBounds(videoEl.clientWidth, videoEl.clientHeight);
        }
    }

    handleVideoPlaying() {
        const { videoStarted } = this.state;

        // Firefox will send multiple events
        if (videoStarted) {
            return;
        }	

        this.context.videoStart();

        this.setState({
            videoStarted: true
        });
    }

    updateVideoStream() {
        const videoEl = this.videoRef.current;

        if (videoEl) {
            this.props.setVideoBounds(videoEl.clientWidth, videoEl.clientHeight);

            if (videoEl.srcObject === null) {
                videoEl.srcObject = this.context.personaVideoObject;
            }
        }
    }

    render() {
        const { isConnected } = this.props;

        return (
            <CSSTransition in={ isConnected } timeout={ 500 } unmountOnExit classNames="video">
                <StyledPersonaVideo { ...this.props }>
                    <video ref={ this.videoRef } onPlaying={ this.handleVideoPlaying } autoPlay playsInline></video>
                </StyledPersonaVideo>
            </CSSTransition>
        );
    }
}

const StyledPersonaVideo = styled.div`
    bottom: ${props => props.isTranscriptOpen ? props.theme.splitHeight : 0};
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transition: opacity 0.5s ease-in-out;
    z-index: 0;

    video {
        width: 100%;
        height: 100%;
    }

    ${Media.tablet`
        bottom: 0;
    `}

    /* Enter/Exit animation */
    &.video-enter {
        opacity: 0;
    }

    &.video-enter-done,
    &.video-enter-active {
        opacity: 1;
    }

    &.video-exit {
        opacity: 1;
    }

    &.video-exit-done,
    &.video-exit-active {
        opacity: 0;
    }
`;

function mapStateToProps(state) {
    return {
        isTranscriptOpen: state.isTranscriptOpen,
        isConnected: state.isConnected,
        personaState: state.personaState
    };
}

function mapDispatchToProps(dispatch) {
    return {
        setVideoBounds: (width, height) => dispatch({
            type: ActionTypes.SET_VIDEO_BOUNDS,
            width: width,
            height: height
        })
    };
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(PersonaVideo);


