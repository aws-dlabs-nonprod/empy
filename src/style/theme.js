export default {
    textPrimary: '#2D373E',
    textSecondary: '#E7E6E6',
    textBody: '#1A1A1A',
    textDark: '#621A4B',

    colourPrimary: '#D5002B',
    colourPrimaryHover: '#e24d6b',
    colourPrimaryTransparent: 'rgba(0, 174, 236, 0.5)',

    colourForeground: '#FFFFFF',
    colourForegroundHover: '#F9F9F8',
    colourForegroundTransparent: 'rgba(255, 255, 255, 0.2)',

    colourInactive: '#7F7F7F',
    colourBackground: '#E5E5E5',
    colourDivider: '#D7D2CB',
    colourFaintBorder: '#8C9296',

    radius: '0.3rem',
    splitHeight: '50vh'
};
