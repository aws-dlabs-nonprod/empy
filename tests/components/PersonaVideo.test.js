import React from 'react';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import 'jest-styled-components';
import theme from '@style/theme';

import PersonaVideo from '@components/PersonaVideo';

const mockStore = configureStore([]);

describe('PersonaVideo', () => {

    test('renders correctly', () => {
        const store = mockStore({});

        const component = mount(
            <Provider store={store}>
                <PersonaVideo theme={ theme } />
            </Provider>
        );

        expect(component.find(PersonaVideo)).toMatchSnapshot();
    });

    test('handles connection', () => {
        const store = mockStore({
            isConnected: true
        });

        const component = mount(
            <Provider store={store}>
                <PersonaVideo theme={ theme } />
            </Provider>
        );

        expect(component.find(PersonaVideo)).toMatchSnapshot();
    });
});
