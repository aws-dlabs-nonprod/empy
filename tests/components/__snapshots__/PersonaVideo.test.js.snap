// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonaVideo handles connection 1`] = `
.c0 {
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
  -webkit-transition: opacity 0.5s ease-in-out;
  transition: opacity 0.5s ease-in-out;
  z-index: 0;
}

.c0 video {
  width: 100%;
  height: 100%;
}

.c0.video-enter {
  opacity: 0;
}

.c0.video-enter-done,
.c0.video-enter-active {
  opacity: 1;
}

.c0.video-exit {
  opacity: 1;
}

.c0.video-exit-done,
.c0.video-exit-active {
  opacity: 0;
}

@media (min-width:768px) and (min-height:0px) {
  .c0 {
    bottom: 0;
  }
}

<Connect(PersonaVideo)
  theme={
    Object {
      "colourBackground": "#E5E5E5",
      "colourDivider": "#D7D2CB",
      "colourFaintBorder": "#8C9296",
      "colourForeground": "#FFFFFF",
      "colourForegroundHover": "#F9F9F8",
      "colourForegroundTransparent": "rgba(255, 255, 255, 0.2)",
      "colourInactive": "#7F7F7F",
      "colourPrimary": "#D5002B",
      "colourPrimaryHover": "#e24d6b",
      "colourPrimaryTransparent": "rgba(0, 174, 236, 0.5)",
      "radius": "0.3rem",
      "splitHeight": "50vh",
      "textBody": "#1A1A1A",
      "textDark": "#621A4B",
      "textPrimary": "#2D373E",
      "textSecondary": "#E7E6E6",
    }
  }
>
  <PersonaVideo
    isConnected={true}
    setVideoBounds={[Function]}
    theme={
      Object {
        "colourBackground": "#E5E5E5",
        "colourDivider": "#D7D2CB",
        "colourFaintBorder": "#8C9296",
        "colourForeground": "#FFFFFF",
        "colourForegroundHover": "#F9F9F8",
        "colourForegroundTransparent": "rgba(255, 255, 255, 0.2)",
        "colourInactive": "#7F7F7F",
        "colourPrimary": "#D5002B",
        "colourPrimaryHover": "#e24d6b",
        "colourPrimaryTransparent": "rgba(0, 174, 236, 0.5)",
        "radius": "0.3rem",
        "splitHeight": "50vh",
        "textBody": "#1A1A1A",
        "textDark": "#621A4B",
        "textPrimary": "#2D373E",
        "textSecondary": "#E7E6E6",
      }
    }
  >
    <CSSTransition
      classNames="video"
      in={true}
      timeout={500}
      unmountOnExit={true}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={500}
        unmountOnExit={true}
      >
        <PersonaVideo__StyledPersonaVideo
          isConnected={true}
          setVideoBounds={[Function]}
          theme={
            Object {
              "colourBackground": "#E5E5E5",
              "colourDivider": "#D7D2CB",
              "colourFaintBorder": "#8C9296",
              "colourForeground": "#FFFFFF",
              "colourForegroundHover": "#F9F9F8",
              "colourForegroundTransparent": "rgba(255, 255, 255, 0.2)",
              "colourInactive": "#7F7F7F",
              "colourPrimary": "#D5002B",
              "colourPrimaryHover": "#e24d6b",
              "colourPrimaryTransparent": "rgba(0, 174, 236, 0.5)",
              "radius": "0.3rem",
              "splitHeight": "50vh",
              "textBody": "#1A1A1A",
              "textDark": "#621A4B",
              "textPrimary": "#2D373E",
              "textSecondary": "#E7E6E6",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "PersonaVideo__StyledPersonaVideo-sc-1fqmpy8-0",
                  "isStatic": false,
                  "lastClassName": "c0",
                  "rules": Array [
                    "bottom:",
                    [Function],
                    ";left:0;position:absolute;right:0;top:0;transition:opacity 0.5s ease-in-out;z-index:0;video{width:100%;height:100%;}",
                    "@media (min-width:",
                    "768",
                    "px) and (min-height:",
                    "0",
                    "px){",
                    "
        bottom: 0;
    ",
                    "}",
                    " &.video-enter{opacity:0;}&.video-enter-done,&.video-enter-active{opacity:1;}&.video-exit{opacity:1;}&.video-exit-done,&.video-exit-active{opacity:0;}",
                  ],
                },
                "displayName": "PersonaVideo__StyledPersonaVideo",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "PersonaVideo__StyledPersonaVideo-sc-1fqmpy8-0",
                "target": "div",
                "toString": [Function],
                "usesTheme": true,
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            isConnected={true}
            setVideoBounds={[Function]}
            theme={
              Object {
                "colourBackground": "#E5E5E5",
                "colourDivider": "#D7D2CB",
                "colourFaintBorder": "#8C9296",
                "colourForeground": "#FFFFFF",
                "colourForegroundHover": "#F9F9F8",
                "colourForegroundTransparent": "rgba(255, 255, 255, 0.2)",
                "colourInactive": "#7F7F7F",
                "colourPrimary": "#D5002B",
                "colourPrimaryHover": "#e24d6b",
                "colourPrimaryTransparent": "rgba(0, 174, 236, 0.5)",
                "radius": "0.3rem",
                "splitHeight": "50vh",
                "textBody": "#1A1A1A",
                "textDark": "#621A4B",
                "textPrimary": "#2D373E",
                "textSecondary": "#E7E6E6",
              }
            }
          >
            <div
              className="c0"
            >
              <video
                autoPlay={true}
                onPlaying={[Function]}
                playsInline={true}
              />
            </div>
          </StyledComponent>
        </PersonaVideo__StyledPersonaVideo>
      </Transition>
    </CSSTransition>
  </PersonaVideo>
</Connect(PersonaVideo)>
`;

exports[`PersonaVideo renders correctly 1`] = `
<Connect(PersonaVideo)
  theme={
    Object {
      "colourBackground": "#E5E5E5",
      "colourDivider": "#D7D2CB",
      "colourFaintBorder": "#8C9296",
      "colourForeground": "#FFFFFF",
      "colourForegroundHover": "#F9F9F8",
      "colourForegroundTransparent": "rgba(255, 255, 255, 0.2)",
      "colourInactive": "#7F7F7F",
      "colourPrimary": "#D5002B",
      "colourPrimaryHover": "#e24d6b",
      "colourPrimaryTransparent": "rgba(0, 174, 236, 0.5)",
      "radius": "0.3rem",
      "splitHeight": "50vh",
      "textBody": "#1A1A1A",
      "textDark": "#621A4B",
      "textPrimary": "#2D373E",
      "textSecondary": "#E7E6E6",
    }
  }
>
  <PersonaVideo
    setVideoBounds={[Function]}
    theme={
      Object {
        "colourBackground": "#E5E5E5",
        "colourDivider": "#D7D2CB",
        "colourFaintBorder": "#8C9296",
        "colourForeground": "#FFFFFF",
        "colourForegroundHover": "#F9F9F8",
        "colourForegroundTransparent": "rgba(255, 255, 255, 0.2)",
        "colourInactive": "#7F7F7F",
        "colourPrimary": "#D5002B",
        "colourPrimaryHover": "#e24d6b",
        "colourPrimaryTransparent": "rgba(0, 174, 236, 0.5)",
        "radius": "0.3rem",
        "splitHeight": "50vh",
        "textBody": "#1A1A1A",
        "textDark": "#621A4B",
        "textPrimary": "#2D373E",
        "textSecondary": "#E7E6E6",
      }
    }
  >
    <CSSTransition
      classNames="video"
      timeout={500}
      unmountOnExit={true}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={500}
        unmountOnExit={true}
      />
    </CSSTransition>
  </PersonaVideo>
</Connect(PersonaVideo)>
`;
